frost/src/spatial.rs

```diff
@@ -0,0 +1,476 @@
+use crate::{Error, FrostLatLonElev, FrostLocation, FrostObs};
+use chrono::prelude::*;
+use rove::data_switch::{SpatialCache, Timestamp};
+use rove::pb::GeoPoint;
+
+fn extract_metadata(
+    mut header: serde_json::Value,
+    time: DateTime<Utc>,
+) -> Result<FrostLatLonElev, Error> {
+    let location = header
+        .get_mut("extra")
+        .ok_or(Error::FindLocation(
+            "couldn't find extra in header".to_string(),
+        ))?
+        .take()
+        .get_mut("station")
+        .ok_or(Error::FindLocation(
+            "couldn't find station field in extra".to_string(),
+        ))?
+        .get_mut("location")
+        .ok_or(Error::FindLocation(
+            "couldn't find location field in station".to_string(),
+        ))?
+        .take();
+
+    let loc = serde_json::from_value::<Vec<FrostLocation>>(location)?;
+
+    let lat_lon_elev = loc
+        .into_iter()
+        .find(|l| time > l.from && time < l.to)
+        .ok_or(Error::FindLocation(
+            "couldn't find relevant location for this observation".to_string(),
+        ))?
+        .value;
+
+    Ok(lat_lon_elev)
+}
+
+fn extract_data(
+    mut resp: serde_json::Value,
+    time: DateTime<Utc>,
+) -> Result<Vec<(FrostObs, FrostLatLonElev)>, Error> {
+    let ts_portion: &mut Vec<serde_json::Value> = resp
+        .get_mut("data")
+        .ok_or(Error::FindObs(
+            "couldn't find data field on root".to_string(),
+        ))?
+        .get_mut("tseries")
+        .ok_or(Error::FindObs(
+            "couldn't find tseries field on data".to_string(),
+        ))?
+        .as_array_mut()
+        .take()
+        .ok_or(Error::FindObs("couldn't get array of tseries".to_string()))?;
+
+    let observations = ts_portion
+        .iter_mut()
+        .map(|ts| {
+            let obs: FrostObs = serde_json::from_value(
+                ts.get_mut("observations")
+                    .ok_or(Error::FindObs(
+                        "couldn't find observations field on tseries".to_string(),
+                    ))?
+                    .get_mut(0)
+                    .ok_or(Error::FindObs("couldn't find first obs".to_string()))?
+                    .take(),
+            )?;
+            let lat_lon_elev = extract_metadata(
+                ts.get_mut("header")
+                    .ok_or(Error::FindObs(
+                        "couldn't find header field on tseries".to_string(),
+                    ))?
+                    .take(),
+                time,
+            )?;
+            Ok((obs, lat_lon_elev))
+        })
+        .collect::<Result<Vec<(FrostObs, FrostLatLonElev)>, Error>>()?;
+
+    Ok(observations)
+}
+
+fn json_to_spatial_cache(
+    resp: serde_json::Value,
+    _polygon: Vec<GeoPoint>,
+    _element: &str,
+    timestamp: DateTime<Utc>,
```

Comment:
Do we need to know the polygon and / or the element for the caching?

Reply:
No, they can be removed here